### ISSUE TITLE
Implement framemarking draft

### ISSUE THREAD

**Issue by ibc** [versatica/mediasoup#298]
* Spec: [Frame Marking RTP Header Extension](https://tools.ietf.org/html/draft-ietf-avtext-framemarking)

Chrome has recently included [temporal layers in H264](https://bugs.chromium.org/p/webrtc/issues/detail_ezt?id=10579) and it seems that using frame marking is the only way to know which temporal layer each RTP packet belongs to.

So, the `EncodingContext` should be also ready to read the Frame Marking RTP Header Extension and fill its layers information with it.

**Comment by ibc** [comment]
NOTE: the framemarking extension must be present when forwarding the packet on a pipe transport, so we may want to always keep it in `Producer::MangleRtpPacket()`.

**Comment by ibc** [comment]
Mostly done in a separate branch (to be pushed). BTW Chrome 74.0.3729.157 (stable) already uses frame-marking for H264 layer IDs signaling:

```
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:1, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:0]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:0, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:31]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:31]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:1, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:31]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:0, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:1, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:0, independent:0, discardable:0, base:0, tid:0, lid:0, tl0picidx:1]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:0, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:31]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:1, independent:0, discardable:0, base:0, tid:0, lid:0, tl0picidx:1]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:1, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:31]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:0, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:1, independent:0, discardable:0, base:1, tid:1, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:0, independent:0, discardable:0, base:0, tid:0, lid:0, tl0picidx:32]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:0, tid:0, lid:0, tl0picidx:32]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:0, tid:0, lid:0, tl0picidx:32]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:1, independent:0, discardable:0, base:0, tid:0, lid:0, tl0picidx:32]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:1, independent:0, discardable:0, base:1, tid:2, lid:0, tl0picidx:1]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:1, end:0, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:147]
RTC::Codecs::H264::ProcessRtpPacket() | framemarking [len:3, start:0, end:0, independent:0, discardable:0, base:0, tid:2, lid:0, tl0picidx:147]
```

**Comment by ibc** [comment]
@jmillan, the frame-marking extension parsing is done. May you add the missing logic into `H264::ProcessRtpPacket()` to set the temporal layer Id and so on? (see usage [here](https://github.com/versatica/mediasoup/blob/frame-marking/worker/src/RTC/Producer.cpp?ts=2#L953)). Note that the `PayloadDescriptor` just reads the RTP payload so, somehow, the `tid` information must be given to it by other means. I'd probably change the `H264::Parse()` and `VP8::Parse` to receive the full packet instead of just the payload and its length (so it can also read the parsed RTP extensions). It can internally use `packet->GetPayload()` and `packet->GetPayloadLength()` so not a big deal.

* NOTE 1: Before the packet arrives to `RtpStreamRecv::ReceivePacket()`, the frame-marking extension id has been set into it, so it's safe to use `packet->ReadFrameMarking()` within `H264::ProcessRtpPacket()`.

* NOTE 2: mediasoup-client 3.0.1 already announces L1T3 for H264 so by updating the mediasoup-demo app you can test it.

**Comment by ibc** [comment]
*Regarding key frame detection via frame marking:*

With framemarking we may also know whether a frame is a keyframe or not (in theory). [Spec](https://tools.ietf.org/html/draft-ietf-avtext-framemarking-07#section-3.2). However it's not 100% true. I've tested by checking `start = 1` and `independent = 1` flags and, while they are true for keyframes (as detected by current `packet->IsKeyFrame()` method) they are also true in many other occasions, specially when the stream starts):

```
// Stream begins:

seq:15007 is start+independent frame +0ms
seq:15007 is key frame +0ms
seq:17811 is start+independent frame +7ms
seq:17811 is key frame +3ms
seq:9583 is start+independent frame +141ms
seq:12016 is start+independent frame +51ms
seq:12034 is start+independent frame +67ms
seq:12035 is start+independent frame +0ms
seq:12036 is start+independent frame +0ms
seq:12067 is start+independent frame +191ms
seq:12075 is start+independent frame +34ms
seq:12081 is start+independent frame +30ms
seq:12086 is start+independent frame +28ms
seq:12087 is start+independent frame +1ms
seq:12090 is start+independent frame +4ms
seq:12093 is start+independent frame +23ms
seq:12094 is start+independent frame +1ms
seq:12103 is start+independent frame +334ms
seq:12106 is start+independent frame +5ms
seq:12108 is start+independent frame +23ms
seq:12109 is start+independent frame +0ms
seq:12110 is start+independent frame +0ms
seq:12118 is start+independent frame +74ms
seq:12122 is start+independent frame +27ms
seq:12126 is start+independent frame +14ms
seq:12127 is start+independent frame +1ms
seq:12130 is start+independent frame +16ms
seq:12131 is start+independent frame +0ms
seq:12139 is start+independent frame +73ms
seq:12140 is start+independent frame +1ms
seq:12141 is start+independent frame +0ms
seq:12144 is start+independent frame +7ms
seq:12150 is start+independent frame +64ms
seq:12153 is start+independent frame +5ms
seq:12154 is start+independent frame +0ms
seq:12157 is start+independent frame +33ms
seq:12160 is start+independent frame +46ms
seq:12161 is start+independent frame +0ms
seq:12168 is start+independent frame +38ms
seq:12169 is start+independent frame +0ms
seq:12175 is start+independent frame +18ms
seq:12176 is start+independent frame +1ms
seq:12179 is start+independent frame +109ms
seq:12189 is start+independent frame +72ms
seq:12196 is start+independent frame +73ms
seq:12204 is start+independent frame +84ms
seq:12213 is start+independent frame +91ms
seq:12217 is start+independent frame +22ms
seq:12218 is start+independent frame +0ms
seq:12238 is start+independent frame +451ms
seq:12239 is start+independent frame +0ms
seq:12246 is start+independent frame +199ms
seq:12247 is start+independent frame +1ms
seq:12270 is start+independent frame +270ms
seq:12271 is start+independent frame +0ms
seq:12272 is start+independent frame +0ms
seq:12283 is start+independent frame +249ms
seq:12286 is start+independent frame +5ms
seq:12289 is start+independent frame +7ms
seq:12297 is start+independent frame +97ms
seq:12303 is start+independent frame +23ms
seq:12304 is start+independent frame +0ms
seq:12307 is start+independent frame +5ms
seq:12310 is start+independent frame +6ms
seq:12311 is start+independent frame +0ms
seq:12314 is start+independent frame +6ms
seq:12321 is start+independent frame +34ms
seq:12329 is start+independent frame +211ms
seq:12330 is start+independent frame +1ms
seq:12333 is start+independent frame +5ms
seq:12336 is start+independent frame +6ms
seq:12337 is start+independent frame +0ms
seq:12338 is start+independent frame +1ms
seq:12341 is start+independent frame +5ms
seq:12344 is start+independent frame +38ms
seq:12347 is start+independent frame +19ms
seq:12348 is start+independent frame +1ms
seq:12357 is start+independent frame +45ms
seq:12360 is start+independent frame +14ms
seq:12363 is start+independent frame +5ms
seq:12364 is start+independent frame +2ms
seq:12365 is start+independent frame +0ms
seq:12377 is start+independent frame +85ms
seq:12386 is start+independent frame +116ms
seq:12387 is start+independent frame +1ms
seq:12398 is start+independent frame +61ms
seq:12399 is start+independent frame +1ms
seq:12406 is start+independent frame +105ms
seq:12409 is start+independent frame +12ms
seq:12410 is start+independent frame +0ms
seq:12421 is start+independent frame +51ms
seq:12429 is start+independent frame +96ms
seq:12430 is start+independent frame +0ms
seq:12431 is start+independent frame +1ms
seq:12447 is start+independent frame +188ms
seq:12466 is start+independent frame +144ms
seq:12490 is start+independent frame +320ms
seq:12537 is start+independent frame +334ms
seq:12538 is start+independent frame +0ms
seq:12549 is start+independent frame +48ms
seq:12550 is start+independent frame +0ms
seq:12551 is start+independent frame +1ms
seq:12579 is start+independent frame +145ms
seq:12580 is start+independent frame +1ms
seq:12605 is start+independent frame +125ms
seq:12614 is start+independent frame +144ms
seq:29220 is start+independent frame +73ms

// Consumer connects (so PLI sent):

seq:29220 is key frame +0ms
seq:17974 is start+independent frame +28ms
seq:17974 is key frame +2ms
seq:15279 is start+independent frame +1ms
seq:15279 is key frame +0ms
seq:19188 is start+independent frame +296ms
seq:19189 is start+independent frame +0ms
seq:19190 is start+independent frame +0ms
seq:19191 is start+independent frame +0ms
seq:19192 is start+independent frame +1ms
seq:19193 is start+independent frame +0ms
seq:9829 is start+independent frame +153ms

// PLIs sent:

seq:31002 is start+independent frame +13s
seq:31002 is key frame +0ms
seq:15908 is start+independent frame +16ms
seq:15908 is key frame +1ms
seq:18332 is start+independent frame +11ms
seq:18332 is key frame +1ms
seq:31156 is start+independent frame +1s
seq:31156 is key frame +1ms
seq:15954 is start+independent frame +11ms
seq:15954 is key frame +0ms
seq:18358 is start+independent frame +15ms
seq:18358 is key frame +0ms
```

Oh **good news!!** All those start+independent packets at the beginning are RTX probation packets (that are discarded by the `RtpStreamRecv` as usual, so this may just work!

**Comment by ibc** [comment]
I've [made](https://github.com/versatica/mediasoup/commit/8aaead55c108631f5e8eab9f7d46ce0da19d35ca) both `H264::Parse()` and `VP8::Parse()` frame-marking aware (just used in H264 for detecting key frame for now).

**Comment by ibc** [comment]
This is done in #305 and released mediasoup 3.0.5 (since this is an important feature).

However there are some TODO comments in `H264::PayloadDescriptorHandler::Encode()` to be checked, so let's keep this issue open for a while. Those TODOs are:

* We should check incremental "picture id" somehow and let it pass the packet if so. However no idea how to check that in H264 frame-marking.
* When in the lowest simulcast stream, there are no H264 packets with the `framingMarking->base` (Base Layer Sync) set, so we cannot wait for it in order to let higher temporal layers pass.
   * We may need [The Layer Refresh Request (LRR) RTCP Feedback](https://tools.ietf.org/html/draft-ietf-avtext-lrr) for it, but libwebrtc does not support it.

**Comment by ibc** [comment]
OK, this is mostly done. Let move the pending issues to #306.
